expirationd.lua

```diff
@@ -56,85 +58,90 @@ local constants = {
 -- Task fibers
 -- ------------------------------------------------------------------------- --
 
-local function construct_key(space_id, tuple)
+-- get all fields in primary key(composite possible) from tuple
+local function construct_key(expire_index, tuple)
     return fun.map(
         function(x) return tuple[x.fieldno] end,
-        box.space[space_id].index[0].parts
+        expire_index.parts
     ):totable()
 end
 
+-- do expiration process on tuple
 local function expiration_process(task, tuple)
     task.checked_tuples_count = task.checked_tuples_count + 1
     if task.is_tuple_expired(task.args, tuple) then
         task.expired_tuples_count = task.expired_tuples_count + 1
-        task.process_expired_tuple(task.space_id, task.args, tuple)
+        task.process_expired_tuple(task.space_id, task.args, tuple, task)
     end
 end
 
-local function suspend_basic(scan_space, task, len)
+-- stop for some time
+local function suspend_basic(task, len)
     local delay = (task.tuples_per_iteration * task.full_scan_time)
     delay = math.min(delay / len, task.iteration_delay)
     fiber.sleep(delay)
 end
 
-local function suspend(scan_space, task)
-    local space_len = scan_space:len()
+local function suspend(task)
+    -- Return the number of tuples in the space
+    local space_len = task.expire_index:len()
     if space_len > 0 then
-        suspend_basic(scan_space, task, space_len)
+        suspend_basic(task, space_len)
     end
 end
 
-local function tree_index_iter(scan_space, task)
+-- delete with some suspend and some tuples limit
+local function tree_index_iter(task)
     -- iteration with GT iterator
-    local params = {iterator = 'GT', limit = task.tuples_per_iteration}
+    local params = {iterator = task.iterator, limit = task.tuples_per_iteration}
     local last_id
-    local tuples = scan_space.index[0]:select({}, params)
+    local tuples = task.expire_index:select(task.start_element, params)
     while #tuples > 0 do
         last_id = tuples[#tuples]
         for _, tuple in ipairs(tuples) do
+            if task:stop_iteration() then goto done end
             expiration_process(task, tuple)
         end
-        suspend(scan_space, task)
-        local key = construct_key(scan_space.id, last_id)
-        tuples = scan_space.index[0]:select(key, params)
+        suspend(task)
+        local key = construct_key(task.expire_index, last_id)
+        -- select all greater then last key
+        tuples = task.expire_index:select(key, params)
```

Comment:
Approach with select won't work for non-unique secondary indexes.
See https://github.com/tarantool/tarantool/issues/3898 for details.

Single allowed you option is to use pairs here - https://github.com/tarantool/expirationd/pull/40

Test case is quite simple, real and painful. Insert `tuples_per_iteration + 1` with the same secondary index value. Then if you use 'GT' iterator you skip such records if you use 'GE' iterator it could be an infinite loop.